File: content/problems/2019-the-score-of-students-solving-math-expression/solutions/correct-evaluation-and-interval-set-dp/Solution.java
class Solution {
    public int scoreOfStudents(String s,int[] answers){
        int count=(s.length()+1)/2;
        int[] numbers=new int[count];
        char[] operators=new char[count-1];
        for(int i=0;i<count;i++){
            numbers[i]=s.charAt(2*i)-'0';
            if(i+1<count)operators[i]=s.charAt(2*i+1);
        }
        int correct=0,term=numbers[0];
        for(int i=0;i<operators.length;i++)if(operators[i]=='*')term*=numbers[i+1];
        else{
            correct+=term;
            term=numbers[i+1];
        }
        correct+=term;
        Set<Integer>[][] dp=new Set[count][count];
        for(int i=0;i<count;i++){
            dp[i][i]=new HashSet<>();
            dp[i][i].add(numbers[i]);
        }
        for(int length=2;length<=count;length++)for(int left=0;left+length<=count;left++){
            int right=left+length-1;
            dp[left][right]=new HashSet<>();
            for(int split=left;split<right;split++)for(int a:dp[left][split])for(int b:dp[split+1][right]){
                int value=operators[split]=='+'?a+b:a*b;
                if(value<=1000)dp[left][right].add(value);
            }
        }
        int score=0;
        for(int answer:answers)score+=answer==correct?5:dp[0][count-1].contains(answer)?2:0;
        return score;
    }
}

File: content/problems/2019-the-score-of-students-solving-math-expression/solutions/correct-evaluation-and-interval-set-dp/Solution.kt
class Solution {
    fun scoreOfStudents(s:String,answers:IntArray):Int{
        val count=(s.length+1)/2;
        val numbers=IntArray(count){
            s[2*it]-'0'
        };
        val operators=CharArray(count-1){
            s[2*it+1]
        };
        var correct=0;
        var term=numbers[0];
        for(i in operators.indices)if(operators[i]=='*')term*=numbers[i+1] else{
            correct+=term;
            term=numbers[i+1]
        };
        correct+=term;
        val dp=Array(count){
            arrayOfNulls<MutableSet<Int>>(count)
        };
        for(i in 0 until count)dp[i][i]=hashSetOf(numbers[i]);
        for(length in 2..count)for(left in 0..count-length){
            val right=left+length-1;
            val values=HashSet<Int>();
            for(split in left until right)for(a in dp[left][split]!!)for(b in dp[split+1][right]!!){
                val value=if(operators[split]=='+')a+b else a*b;
                if(value<=1000)values.add(value)
            };
            dp[left][right]=values
        };
        return answers.fold(0){score,answer ->
            score+if(answer==correct)5 else if(answer in dp[0][count-1]!!)2 else 0
        }
    }
}
